File: java/base/src/main/java/vsdk/toolkit/material/PhongBidirectionalTransmittanceDistributionFunction.java
package vsdk.toolkit.material;

/**
Index of refraction data type. Normally when using BSDF's
this should not be needed. In C++ this would of course
be a plain complex number
*/

/**
All components of the Btdf

Vector directions :

in : towards patch
out : from patch
normal : leaving from patch, on the incoming side.
         So in.normal < 0 !!!
*/

/**
Returns the transmittance of the BTDF
*/

/**
Btdf evaluations
*/

// Specular-like refraction can turn into reflection.

// So for refraction a complete sphere should be

// sampled ! Importance sampling is advisable.

// Diffuse transmission is considered to always pass

// the material boundary

// Diffuse part

// Normal is pointing away from refracted direction

// Specular part

// cos(a) ^ n

// Ks -> ks

// Choose sampling mode

// Store in phong data ?

// Determine diffuse or glossy/specular sampling

// Absorption

// Rescaling of x_1

// Sample diffuse

// Section [ARVO1995b].2: square-to-sphere mapping in the frame of the transmitted hemisphere.

// Sample specular

// Section [ARVO1995b].2: same 2D mapping with a lobe centered on the ideal transmitted direction.

// Assume totalIR (maybe we should test the refractionIndices

// Combine Probability Density Functions

// Ensure 'in' on the same side as 'normal'!

// Transmitted ray

// Diffuse sampling probabilityDensityFunction

// Glossy or specular

// Normal was inverted, so materialSides switch also

/**
Refraction index
*/

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.logging.Logger;
import vsdk.toolkit.common.linealAlgebra.CoordinateSystem;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector3D;

public class PhongBidirectionalTransmittanceDistributionFunction {
    private ColorRgb Kd;
    private ColorRgb Ks;
    private float avgKd;
    private float avgKs;
    private float Ns;
    private RefractionIndex refractionIndex;

    private boolean isSpecular() {
        return Ns >= Xxdf.PHONG_LOWEST_SPECULAR_EXP;
    }

    public PhongBidirectionalTransmittanceDistributionFunction(
        ColorRgb inKd,
        ColorRgb inKs,
        float inNs,
        float inNr,
        float inNi) {
        Kd = new ColorRgb(inKd.r, inKd.g, inKd.b);
        avgKd = Kd.average();
        Ks = new ColorRgb(inKs.r, inKs.g, inKs.b);
        avgKs = Ks.average();
        Ns = inNs;
        refractionIndex = new RefractionIndex();
        refractionIndex.set(inNr, inNi);
    }

    public ColorRgb getKd() {
        return Kd;
    }

    public ColorRgb getKs() {
        return Ks;
    }

    public float getNs() {
        return Ns;
    }

    public RefractionIndex getRefractionIndex() {
        return refractionIndex;
    }

    public ColorRgb transmittance(int flags) {
        ColorRgb result = new ColorRgb();
        result.clear();

        if ((flags & XxdfComponentFlag.DIFFUSE_COMPONENT) != 0) {
            result.add(result, Kd);
        }

        if (isSpecular()) {
            if ((flags & XxdfComponentFlag.SPECULAR_COMPONENT) != 0) {
                result.add(result, Ks);
            }
        }
        else if ((flags & XxdfComponentFlag.GLOSSY_COMPONENT) != 0) {
            result.add(result, Ks);
        }

        if (!Float.isFinite(result.average())) {
            Logger.fatal(-1, "transmittance", "Oops - result is not finite!");
        }

        return result;
    }

    public ColorRgb evaluate(
        RefractionIndex inIndex,
        RefractionIndex outIndex,
        Vector3D in,
        Vector3D out,
        Vector3D normal,
        int flags) {
        Vector3D inRev = new Vector3D();
        inRev.scaledCopy(-1.0f, in);

        ColorRgb result = new ColorRgb();
        result.clear();

        if (((flags & XxdfComponentFlag.DIFFUSE_COMPONENT) != 0) && (avgKd > 0.0f)) {
            boolean isReflection = normal.dotProduct(out) >= 0.0f;
            if (!isReflection) {
                result = new ColorRgb(Kd.r, Kd.g, Kd.b);
                result.scale((float)(1.0 / Math.PI));
            }
        }

        int nonDiffuseFlag = isSpecular() ? XxdfComponentFlag.SPECULAR_COMPONENT : XxdfComponentFlag.GLOSSY_COMPONENT;
        if (((flags & nonDiffuseFlag) != 0) && (avgKs > 0.0f)) {
            boolean[] totalIR = new boolean[] {false};
            Vector3D idealRefracted = Xxdf.idealRefractedDirection(inRev, normal, inIndex, outIndex, totalIR);
            float localDotProduct = idealRefracted.dotProduct(out);

            if (localDotProduct > 0.0f) {
                float tmpFloat = (float)Math.pow(localDotProduct, Ns);
                tmpFloat *= (Ns + 2.0f) / (2.0f * (float)Math.PI);
                result.addScaled(result, tmpFloat, Ks);
            }
        }

        return result;
    }

    public Vector3D sample(
        RefractionIndex inIndex,
        RefractionIndex outIndex,
        Vector3D in,
        Vector3D normal,
        int doRussianRoulette,
        int flags,
        double x1,
        double x2,
        double[] probabilityDensityFunction) {
        setOut(probabilityDensityFunction, 0.0);

        Vector3D newDir = new Vector3D(0.0f, 0.0f, 0.0f);
        Vector3D inRev = new Vector3D();
        inRev.scaledCopy(-1.0f, in);

        double localAverageKd = ((flags & XxdfComponentFlag.DIFFUSE_COMPONENT) != 0) ? avgKd : 0.0;
        int nonDiffuseFlag = isSpecular() ? XxdfComponentFlag.SPECULAR_COMPONENT : XxdfComponentFlag.GLOSSY_COMPONENT;
        double localAverageKs = ((flags & nonDiffuseFlag) != 0) ? avgKs : 0.0;

        double scatteredPower = localAverageKd + localAverageKs;
        if (scatteredPower < Numeric.EPSILON) {
            return newDir;
        }

        if (doRussianRoulette != 0) {
            if (x1 > scatteredPower) {
                return newDir;
            }
            x1 /= scatteredPower;
        }

        boolean[] totalIR = new boolean[] {false};
        Vector3D idealDir = Xxdf.idealRefractedDirection(inRev, normal, inIndex, outIndex, totalIR);
        Vector3D invNormal = new Vector3D();
        invNormal.scaledCopy(-1.0f, normal);
        CoordinateSystem coord = new CoordinateSystem();

        double diffPdf;
        double nonDiffPdf;

        if (x1 < (localAverageKd / scatteredPower)) {
            x1 = x1 / (localAverageKd / scatteredPower);
            coord.setFromZAxis(invNormal);
            double[] outDiffPdf = new double[] {0.0};
            newDir = coord.sampleHemisphereCosTheta(x1, x2, outDiffPdf);
            diffPdf = outDiffPdf[0];

            float tmpFloat = idealDir.dotProduct(newDir);
            if (tmpFloat > 0.0f) {
                nonDiffPdf = (Ns + 1.0) * Math.pow(tmpFloat, Ns) / (2.0 * Math.PI);
            }
            else {
                nonDiffPdf = 0.0;
            }
        }
        else {
            x1 = (x1 - (localAverageKd / scatteredPower)) / (localAverageKs / scatteredPower);
            coord.setFromZAxis(idealDir);
            double[] outNonDiffPdf = new double[] {0.0};
            newDir = coord.sampleHemisphereCosNTheta(Ns, x1, x2, outNonDiffPdf);
            nonDiffPdf = outNonDiffPdf[0];

            double cosTheta = normal.dotProduct(newDir);
            if (cosTheta > 0.0) {
                diffPdf = cosTheta / Math.PI;
            }
            else {
                diffPdf = 0.0;
            }
        }

        double pdf = localAverageKd * diffPdf + localAverageKs * nonDiffPdf;
        if (doRussianRoulette == 0) {
            pdf /= scatteredPower;
        }

        setOut(probabilityDensityFunction, pdf);
        return newDir;
    }

    public void evaluateProbabilityDensityFunction(
        RefractionIndex inIndex,
        RefractionIndex outIndex,
        Vector3D in,
        Vector3D out,
        Vector3D normal,
        int flags,
        double[] probabilityDensityFunction,
        double[] probabilityDensityFunctionRR) {
        setOut(probabilityDensityFunction, 0.0);
        setOut(probabilityDensityFunctionRR, 0.0);

        Vector3D inRev = new Vector3D();
        inRev.scaledCopy(-1.0f, in);

        Vector3D goodNormal = new Vector3D();
        double cosIn = in.dotProduct(normal);
        if (cosIn >= 0.0) {
            goodNormal.copy(normal);
        }
        else {
            goodNormal.scaledCopy(-1.0f, normal);
        }

        double cosTheta = goodNormal.dotProduct(out);

        double localAverageKd;
        if (((flags & XxdfComponentFlag.DIFFUSE_COMPONENT) != 0) && (cosTheta < 0.0)) {
            localAverageKd = avgKd;
        }
        else {
            localAverageKd = 0.0;
        }

        int nonDiffuseFlag = isSpecular() ? XxdfComponentFlag.SPECULAR_COMPONENT : XxdfComponentFlag.GLOSSY_COMPONENT;
        double localAverageKs = ((flags & nonDiffuseFlag) != 0) ? avgKs : 0.0;

        double scatteredPower = localAverageKd + localAverageKs;
        if (scatteredPower < Numeric.EPSILON) {
            return;
        }

        double diffPdf = (localAverageKd > 0.0) ? (cosTheta / Math.PI) : 0.0;

        double nonDiffPdf = 0.0;
        if (localAverageKs > 0.0) {
            boolean[] totalIR = new boolean[] {false};
            Vector3D idealDir;

            if (cosIn >= 0.0) {
                idealDir = Xxdf.idealRefractedDirection(inRev, goodNormal, inIndex, outIndex, totalIR);
            }
            else {
                idealDir = Xxdf.idealRefractedDirection(inRev, goodNormal, outIndex, inIndex, totalIR);
            }

            double cosAlpha = idealDir.dotProduct(out);
            if (cosAlpha > 0.0) {
                nonDiffPdf = (Ns + 1.0) * Math.pow(cosAlpha, Ns) / (2.0 * Math.PI);
            }
        }

        setOut(probabilityDensityFunction, (localAverageKd * diffPdf + localAverageKs * nonDiffPdf) / scatteredPower);
        setOut(probabilityDensityFunctionRR, scatteredPower);
    }

    public void setIndexOfRefraction(RefractionIndex index) {
        index.set(refractionIndex.getNr(), refractionIndex.getNi());
    }

    private static void setOut(double[] out, double value) {
        if (out != null && out.length > 0) {
            out[0] = value;
        }
    }
}

File: cpp/base/src/vsdk/toolkit/material/PhongBidirectionalTransmittanceDistributionFunction.cpp
#include "vsdk/toolkit/java/lang/Float.h"
#include "vsdk/toolkit/common/linealAlgebra/CoordinateSystem.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/material/RendererConfiguration.h"
#include "vsdk/toolkit/material/PhongBidirectionalReflectanceDistributionFunction.h"
#include "vsdk/toolkit/material/PhongBidirectionalTransmittanceDistributionFunction.h"
#include "vsdk/toolkit/material/Xxdf.h"

/**
All components of the Btdf

Vector directions :

in : towards patch
out : from patch
normal : leaving from patch, on the incoming side.
         So in.normal < 0 !!!
*/

bool
PhongBidirectionalTransmittanceDistributionFunction::isSpecular() const {
    return Ns >= Xxdf::PHONG_LOWEST_SPECULAR_EXP;
}

PhongBidirectionalTransmittanceDistributionFunction::PhongBidirectionalTransmittanceDistributionFunction(
    const ColorRgb *inKd, const ColorRgb *inKs, const float inNs, const float inNr, const float inNi):
    Kd(*inKd),
    Ks(*inKs),
    avgKd(Kd.average()),
    avgKs(Ks.average()),
    Ns(inNs),
    refractionIndex()
{
    refractionIndex.set(inNr, inNi);
}

PhongBidirectionalTransmittanceDistributionFunction::~PhongBidirectionalTransmittanceDistributionFunction() {
}

/**
Returns the transmittance of the BTDF
*/
ColorRgb
PhongBidirectionalTransmittanceDistributionFunction::transmittance(char flags) const {
    double r = 0.0;
    double g = 0.0;
    double b = 0.0;

    if ( flags & DIFFUSE_COMPONENT ) {
        r += Kd.getR();
        g += Kd.getG();
        b += Kd.getB();
    }

    if ( isSpecular() ) {
        if ( flags & SPECULAR_COMPONENT ) {
            r += Ks.getR();
            g += Ks.getG();
            b += Ks.getB();
        }
    } else {
        if ( flags & GLOSSY_COMPONENT ) {
            r += Ks.getR();
            g += Ks.getG();
            b += Ks.getB();
        }
    }

    const ColorRgb result(r, g, b);
    if ( !java::Float::isFinite(static_cast<float>(result.average())) ) {
        Logger::fatal(-1, "transmittance", "Oops - result is not finite!");
    }

    return result;
}

/**
Btdf evaluations
*/
ColorRgb
PhongBidirectionalTransmittanceDistributionFunction::evaluate(
    RefractionIndex inIndex,
    RefractionIndex outIndex,
    const Vector3D *in,
    const Vector3D *out,
    const Vector3D *normal,
    char flags) const
{
    Vector3D inRev;
    inRev.scaledCopy(-1.0, *in);

    // Specular-like refraction can turn into reflection.
    // So for refraction a complete sphere should be
    // sampled ! Importance sampling is advisable.
    // Diffuse transmission is considered to always pass
    // the material boundary
    double r = 0.0;
    double g = 0.0;
    double b = 0.0;

    if ( (flags & DIFFUSE_COMPONENT) && (avgKd > 0) ) {
        // Diffuse part

        // Normal is pointing away from refracted direction
        bool isReflection = (normal->dotProduct(*out) >= 0);

        if ( !isReflection ) {
            r += M_1_PI * Kd.getR();
            g += M_1_PI * Kd.getG();
            b += M_1_PI * Kd.getB();
        }
    }

    char nonDiffuseFlag;

    if ( isSpecular() ) {
        nonDiffuseFlag = SPECULAR_COMPONENT;
    } else {
        nonDiffuseFlag = GLOSSY_COMPONENT;
    }

    if ( (flags & nonDiffuseFlag) && (avgKs > 0) ) {
        // Specular part
        bool totalIR;
        Vector3D idealRefracted = Xxdf::idealRefractedDirection(&inRev, normal, inIndex, outIndex, &totalIR);
        float localDotProduct = idealRefracted.dotProduct(*out);

        if ( localDotProduct > 0 ) {
            float tmpFloat = java::Math::pow(localDotProduct, Ns); // cos(a) ^ n
            tmpFloat *= (Ns + 2.0F) / (2.0F * static_cast<float>(M_PI)); // Ks -> ks
            r += tmpFloat * Ks.getR();
            g += tmpFloat * Ks.getG();
            b += tmpFloat * Ks.getB();
        }
    }

    return {r, g, b};
}

Vector3D
PhongBidirectionalTransmittanceDistributionFunction::sample(
    RefractionIndex inIndex,
    RefractionIndex outIndex,
    const Vector3D *in,
    const Vector3D *normal,
    int doRussianRoulette,
    char flags,
    double x1,
    double x2,
    double *probabilityDensityFunction) const
{
    Vector3D newDir = {0.0, 0.0, 0.0};
    bool totalIR;
    Vector3D idealDir;
    Vector3D invNormal;
    CoordinateSystem coord;
    double cosTheta;
    double localAverageKd;
    double localAverageKs;
    double scatteredPower;
    double diffPdf;
    double nonDiffPdf;
    float tmpFloat;
    char nonDiffuseFlag;
    Vector3D inRev;
    inRev.scaledCopy(-1.0, *in);

    *probabilityDensityFunction = 0;

    // Choose sampling mode
    if ( flags & DIFFUSE_COMPONENT ) {
        localAverageKd = avgKd; // Store in phong data ?
    } else {
        localAverageKd = 0.0;
    }

    if ( isSpecular() ) {
        nonDiffuseFlag = SPECULAR_COMPONENT;
    } else {
        nonDiffuseFlag = GLOSSY_COMPONENT;
    }

    if ( flags & nonDiffuseFlag ) {
        localAverageKs = avgKs;
    } else {
        localAverageKs = 0.0;
    }

    scatteredPower = localAverageKd + localAverageKs;

    if ( scatteredPower < Numeric::EPSILON ) {
        return newDir;
    }

    // Determine diffuse or glossy/specular sampling
    if ( doRussianRoulette ) {
        if ( x1 > scatteredPower ) {
            // Absorption
            return newDir;
        }

        // Rescaling of x_1
        x1 /= scatteredPower;
    }

    idealDir = Xxdf::idealRefractedDirection(&inRev, normal, inIndex, outIndex, &totalIR);
    invNormal.scaledCopy(-1, *normal);

    if ( x1 < (localAverageKd / scatteredPower) ) {
        // Sample diffuse
        x1 = x1 / (localAverageKd / scatteredPower);

        // Section [ARVO1995b].2: square-to-sphere mapping in the frame of the transmitted hemisphere.
        coord.setFromZAxis(&invNormal);

        newDir = coord.sampleHemisphereCosTheta(x1, x2, &diffPdf);

        tmpFloat = idealDir.dotProduct(newDir);

        if ( tmpFloat > 0 ) {
            nonDiffPdf = (Ns + 1.0) * java::Math::pow(tmpFloat, Ns) / (2.0 * M_PI);
        } else {
            nonDiffPdf = 0;
        }
    } else {
        // Sample specular
        x1 = (x1 - (localAverageKd / scatteredPower)) / (localAverageKs / scatteredPower);

        // Section [ARVO1995b].2: same 2D mapping with a lobe centered on the ideal transmitted direction.
        coord.setFromZAxis(&idealDir);
        newDir = coord.sampleHemisphereCosNTheta(Ns, x1, x2, &nonDiffPdf);

        cosTheta = normal->dotProduct(newDir);
        if ( cosTheta > 0 ) {
            diffPdf = cosTheta / M_PI;
        } else {
            // Assume totalIR (maybe we should test the refractionIndices
            diffPdf = 0.0;
        }
    }

    // Combine Probability Density Functions
    *probabilityDensityFunction = localAverageKd * diffPdf + localAverageKs * nonDiffPdf;

    if ( !doRussianRoulette ) {
        *probabilityDensityFunction /= scatteredPower;
    }

    return newDir;
}

void
PhongBidirectionalTransmittanceDistributionFunction::evaluateProbabilityDensityFunction(
    RefractionIndex inIndex,
    RefractionIndex outIndex,
    const Vector3D *in,
    const Vector3D *out,
    const Vector3D *normal,
    char flags,
    double *probabilityDensityFunction,
    double *probabilityDensityFunctionRR) const
{
    Vector3D inRev;
    inRev.scaledCopy(-1.0, *in);

    *probabilityDensityFunction = 0;
    *probabilityDensityFunctionRR = 0;

    // Ensure 'in' on the same side as 'normal'!
    Vector3D goodNormal;
    double cosIn = in->dotProduct(*normal);

    if ( cosIn >= 0 ) {
        goodNormal.copy(*normal);
    } else {
        goodNormal.scaledCopy(-1, *normal);
    }

    double cosTheta = goodNormal.dotProduct(*out);

    double localAverageKd;
    if ( flags & DIFFUSE_COMPONENT && (cosTheta < 0) ) {
        // Transmitted ray
        localAverageKd = avgKd;
    } else {
        localAverageKd = 0.0;
    }

    char nonDiffuseFlag;
    if ( isSpecular() ) {
        nonDiffuseFlag = SPECULAR_COMPONENT;
    } else {
        nonDiffuseFlag = GLOSSY_COMPONENT;
    }

    double localAverageKs;
    if ( flags & nonDiffuseFlag ) {
        localAverageKs = avgKs;
    } else {
        localAverageKs = 0.0;
    }

    double scatteredPower = localAverageKd + localAverageKs;

    if ( scatteredPower < Numeric::EPSILON ) {
        return;
    }

    // Diffuse sampling probabilityDensityFunction
    double diffPdf;
    if ( localAverageKd > 0 ) {
        diffPdf = cosTheta / M_PI;
    } else {
        diffPdf = 0.0;
    }

    // Glossy or specular
    double nonDiffPdf = 0.0;
    if ( localAverageKs > 0 ) {
        double cosAlpha;
        Vector3D idealDir;
        bool totalIR;

        if ( cosIn >= 0 ) {
            idealDir = Xxdf::idealRefractedDirection(&inRev, &goodNormal, inIndex,
                                               outIndex, &totalIR);
        } else {
            // Normal was inverted, so materialSides switch also
            idealDir = Xxdf::idealRefractedDirection(&inRev, &goodNormal, outIndex,
                                               inIndex, &totalIR);
        }

        cosAlpha = idealDir.dotProduct(*out);

        nonDiffPdf = 0.0;
        if ( cosAlpha > 0 ) {
            nonDiffPdf = (Ns + 1.0) * java::Math::pow(cosAlpha, static_cast<double>(Ns)) / (2.0 * M_PI);
        }
    }

    *probabilityDensityFunction = (localAverageKd * diffPdf + localAverageKs * nonDiffPdf) / scatteredPower;
    *probabilityDensityFunctionRR = scatteredPower;
}

#ifdef RAYTRACING_ENABLED
/**
Refraction index
*/
void
PhongBidirectionalTransmittanceDistributionFunction::setIndexOfRefraction(RefractionIndex *index) const {
    index->set(refractionIndex.getNr(), refractionIndex.getNi());
}
#endif
